src/pymor/reductors/h2.py

```diff
@@ -644,6 +644,201 @@ def reconstruct(self, u):
+               projection='orth'):
+        r"""Reduce using gap-IRKA.
+
+        See [BBG19]_ Algorithm 1.
+
+        Parameters
+        ----------
+        rom0_params
+            Can be:
+
+            - order of the reduced model (a positive integer),
+            - initial interpolation points (a 1D |NumPy array|),
+            - dict with `'sigma'`, `'b'`, `'c'` as keys mapping to
+              initial interpolation points (a 1D |NumPy array|), right
+              tangential directions (|VectorArray| from
+              `fom.input_space`), and left tangential directions
+              (|VectorArray| from `fom.output_space`), all of the same
+              length (the order of the reduced model),
+            - initial reduced-order model (|LTIModel|).
+
+            If the order of reduced model is given, initial
+            interpolation data is generated randomly.
+        tol
+            Tolerance for the convergence criterion.
+        maxit
+            Maximum number of iterations.
+        num_prev
+            Number of previous iterations to compare the current
+            iteration to. A larger number can avoid occasional cyclic
+            behavior.
+        conv_crit
+            Convergence criterion:
+
+            - `'sigma'`: relative change in interpolation points
+            - `'htwogap'`: :math:`\mathcal{H}_2-gap` distance of
+              reduced-order models divided by :math:`\mathcal{L}_2`
+              norm of new reduced-order model
+            - `'ltwo'`: relative :math:`\mathcal{L}_2` distance of
+              reduced-order models
+        projection
+            Projection method:
+
+            - `'orth'`: projection matrix is orthogonalized with respect
+              to the Euclidean inner product,
+            - `'Eorth'`: projection matrix is orthogonalized with
+              respect to the E product.
+
+        Returns
+        -------
+        rom
+            Reduced |LTIModel| model.
+        """
+        if not self.fom.cont_time:
+            raise NotImplementedError
+
+        self._clear_lists()
+        sigma, b, c = self._rom0_params_to_sigma_b_c(rom0_params)
+        self._store_sigma_b_c(sigma, b, c)
+        assert projection in ('orth', 'Eorth')
+
+        self.logger.info('Starting gap IRKA')
+        self._conv_data = (num_prev + 1) * [None]
+        if conv_crit == 'sigma':
+            self._conv_data[0] = sigma
+        for it in range(maxit):
+            self._set_V_W_reductor(sigma, b, c, projection)
+            rom = self._pg_reductor.reduce()
+            sigma, b, c, gap_rom = self._unstable_rom_to_sigma_b_c(rom)
+            self._store_sigma_b_c(sigma, b, c)
+            self._update_conv_data(sigma, gap_rom, rom, conv_crit)
+            self._compute_conv_crit(rom, conv_crit, it)
+            if self.conv_crit[-1] < tol:
+                break
+
+        return rom
+
+    def _rom0_params_to_sigma_b_c(self, rom0_params):
```

Comment:
It seems unfortunate to copy large parts of `GenericIRKAReductor`. Would it be possible to make `GenericIRKAReductor` even more generic and reduce copies here?

Reply:
I could give the `_rom_to_sigma_b_c` method of `GenericIRKAReductor` an extra output ( i.e. change `return sigma, b, c` to `return sigma, b, c, None`). The `None` output can then be changed to some relevant value in specialized classes, e.g. in `GapIRKAReductor` the `None` would become `gap_rom`. This way I could eventually completely remove the `_rom0_params_to_sigma_b_c` method from `GapIRKAReductor`, is this what you mean by more generic?

Reply:
Hmm, now that I looked more closely at other methods you needed to change, maybe it would be better to do the refactoring in another PR and focus here on just adding the methods. I'm not completely happy with the state of H2-type methods and this new method is good motivation to re-think the structure.